### PR TITLE
a proposal for dumping the info in fwk jobreport

### PR DESCRIPTION
**Pull request by vlimant**
#### PR description:

While looking into #39197 it seems that CondDBESSource dumps in cout, which is, as a comment in the code says "not optimal".
I propose to dump the information in the framework job report with this PR

#### PR validation:

`runTheMatrix.py --what nano --site "" --command '--number 5 --customise_commands "process.GlobalTag.DumpStat=cms.untracked.bool(True)"' -l 2500.332 --job-report`

and I was able to see the entries in the job report.

### CODE REVIEW

**Comment by vlimant** [comment]
assign core

File: CondCore/ESSources/plugins/CondDBESSource.cc [hunk 1]
```diff
@@ -323,7 +325,37 @@ void CondDBESSource::fillList(const std::string& stringList,
...
+    std::map<std::string, std::string> db_info;
+    db_info["DataProxy"] = std::to_string(m_stats.nData);
+    db_info["SetInterval"] = std::to_string(m_stats.nSet);
+    db_info["Runs"] = std::to_string(m_stats.nRun);
+    db_info["Lumis"] = std::to_string(m_stats.nLumi);
+    db_info["Refresh"] = std::to_string(m_stats.nRefresh);
+    db_info["ActualRefresh"] = std::to_string(m_stats.nActualRefresh);
+    db_info["Reconnect"] = std::to_string(m_stats.nReconnect);
+    db_info["ActualReconnect"] = std::to_string(m_stats.nActualReconnect);
+    reportSvc->reportPerformanceSummary("CondDBESSourceStats", db_info);
```
Comment (francescobrivio) [inline CondCore/ESSources/plugins/CondDBESSource.cc:356]: Hi @vlimant!
Sorry for the maybe naive question, but wouldn't it be enough to change the `cout`s to the recommended `edm::LogInfo` or `edm::LogDebug`?

File: CondCore/ESSources/plugins/CondDBESSource.cc [hunk 2]
```diff
@@ -323,7 +325,37 @@ void CondDBESSource::fillList(const std::string& stringList,
...
+  }
+
```
Comment (francescobrivio) [inline CondCore/ESSources/plugins/CondDBESSource.cc:358]: Also I believe you need to also remove the lines below which actually contain the `cout`.

**Comment by makortel** [comment]
Could you elaborate the use case? What is the benefit of using the job report? Who would be reading (and when) the information from the xml file?

**Comment by vlimant** [comment]
@makortel ; same as for time service or other clients reporting to the jobreport occasionally. This can be a way to get the information parsed programatically downstream.
I'd love to hear if there was another simpler way

File: CondCore/ESSources/plugins/CondDBESSource.cc (hunk 1 above)
Comment (vlimant) [reply CondCore/ESSources/plugins/CondDBESSource.cc:356]: Hello @francescobrivio , that is an alternative, though it still requires to parse the log in a non-trivial way ; not like the xml of job report.
any other way would be good for me though

File: CondCore/ESSources/plugins/CondDBESSource.cc (hunk 2 above)
Comment (vlimant) [reply CondCore/ESSources/plugins/CondDBESSource.cc:358]: true ;-)

**Comment by makortel** [comment]
> same as for time service or other clients reporting to the jobreport occasionally. This can be a way to get the information parsed programatically downstream.

Right, but is someone actually going to parse the information programmatically?

**Comment by vlimant** [comment]
do you mean in a production workflow tool of some sort? 
having a systematic check on GT dependency during MINI/NANO integration would be a plus and definitely would require the ability for parsing.
passed this question ; is there a better way to report in fwk job report (or somewhere else)? should a reportConditionUsage be added with adequate formatting and node naming ?

File: CondCore/ESSources/plugins/CondDBESSource.cc (hunk 1 above)
Comment (tvami) [reply CondCore/ESSources/plugins/CondDBESSource.cc:356]: but `edm::LogPrint()` would not require any special parsing, right?

Comment (vlimant) [reply CondCore/ESSources/plugins/CondDBESSource.cc:356]: LogPrint or LogDebug would still require a bunch of grep and else to be able to put this into usable form.
an alternative would be to write out a json directly from CondDBESSource when a dump_stats is requested.
I don't really understand why not using an existing vessel of framework information

**Comment by makortel** [comment]
> do you mean in a production workflow tool of some sort?

I was after for a consumer of the information in any context (be it production, CI, or private tests).

> having a systematic check on GT dependency during MINI/NANO integration would be a plus and definitely would require the ability for parsing.

I agree with that. I also think at least some of the `CondDBESSource` information _might_ be generally interesting to record in monitoring.

But the main intention of the framework job report is mainly to pass information to DMWM. The job report is not intended for storing arbitrary data.

If the only (foreseen) use of this information would be in CI, wouldn't it be simpler to format the data here in any easily-parsable format than going through the hoops of `JobReport`? AlCaDB would have full control of the information and its formatting.

(if there was a use case involving production, we'd need to discuss also enabling the functionality by default)

> passed this question ; is there a better way to report in fwk job report (or somewhere else)? should a reportConditionUsage be added with adequate formatting and node naming ?

I agree, **If** this information would be stored in the framework job report, it would need to be structured properly.

**Comment by vlimant** [comment]
Thanks @makortel for clarifying the main purpose of the fwkjob report. I mostly used it because it has an easy interface (though probably not adapted as is for the payload information) to output an XML. Is there a more general tool in CMSSW to output XML (or json or else) ?

Would @cms-sw/alca-l2 have a better suggestion for a way to make the `CondDBESSourceStats` available ?

**Comment by tvami** [comment]
> Would https://github.com/orgs/cms-sw/teams/alca-l2 have a better suggestion for a way to make the CondDBESSourceStats available ?

@ggovi what's your take of this?

**Comment by ggovi** [comment]
I'm actually a bit surprised about the interest of infos concerning the refresh and reconnect stats. Not sure what type of investigation is targeted here ( at least I can't guess from the above thread ). Maybe JR can explain a little bit what type of data is required here and for what purpose?

**Comment by vlimant** [comment]
@tvami and @cms-sw/xpog-l2 were looking into the GT dependency (or too much thereof) of NANO workflows in #39197 and dumping the information in a parsable way would be beneficial, also in a view to maybe keep a close (and automatic) eye on this with a formal report in PR integration

**Comment by ggovi** [comment]
Ok, so maybe @tvami can provide the list of interesting params in this context? Is it essentially about what record are initiated (not sure it that implies that they are finally consumed )?

**Comment by tvami** [comment]
Let me give a bit more details on what I did in https://github.com/cms-sw/cmssw/issues/39197
I used `process.GlobalTag.DumpStat=cms.untracked.bool(True)` which then gives a huge list, like

```

JetCorrectionsRecord / AK5PFTowers_hiGoodTightTracks: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_AK5PFTowers_hiGoodTightTracks_v0_offline
  
  Time look up, payloadIds:


JetCorrectionsRecord / AK5PFchs: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_CSA14_V4_DATA_AK5PFchs
  
  Time look up, payloadIds:


JetCorrectionsRecord / AK5PFchsHLT: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_AK5PFchs_v1_hlt
  
  Time look up, payloadIds:


JetCorrectionsRecord / AK6PF: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_CSA14_V4_DATA_AK6PF
  
  Time look up, payloadIds:


JetCorrectionsRecord / AK6PFchs: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_CSA14_V4_DATA_AK6PFchs
  
  Time look up, payloadIds:


JetCorrectionsRecord / AK7Calo: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_CSA14_V4_DATA_AK7Calo
  
  Time look up, payloadIds:


JetCorrectionsRecord / AK7PF: frontier://FrontierProd/CMS_CONDITIONS, JetCorrectorParametersCollection_CSA14_V4_DATA_AK7PF
  
  Time look up, payloadIds:
```
which means none of these are consumed.

**Comment by vlimant** [comment]
@ggovi : what should we do here ?

**Comment by vlimant** [comment]
@ggovi : any solution (but parsing a plain text) would work for me at this point

**Comment by tvami** [comment]
@ggovi can you please chime in here?

**Comment by vlimant** [comment]
we will reopen this next time there is an issue with creating a GT for old data in recent release : point at which we will have to have a proper tool to debug this.

**Comment by malbouis** [comment]
Hi all, let me reopen this discussion. We have been looking for the information of consumed conditions quite a lot from PPD side and think it would be useful to have this functionality, be it with the proposed option from @vlimant to dump the information in the framework job report or to have it instead in a json file.
@cms-sw/alca-l2 , @cms-sw/db-l2 , would you please take it from here? Ideally, we should have a nice looking table of the consumed records for a particular GT and step. Or maybe just a list of records that are consumed per GT/step.
@cms-sw/core-l2 , currently this information is being dumped in the log files instead of the framework job report.
@vlimant , I think we would maybe also need the `payloadIds` to be written out. AFAIU, this is how we know if a given payload has been consumed.

**Pull request closed without merging**